Add new StandardMaterial properties to allow users to control FPS-style objects (hands, weapons, tools close to the camera)

clayjohn opened the pull request:
Closes: https://github.com/godotengine/godot-proposals/issues/4443
Closes: https://github.com/godotengine/godot-proposals/issues/11812

For FPS games managing arms/objects/weapons can be a nightmare as you want them to have believable lighting, but you also want to ensure they don't clip through walls and you need to tweak their perspective as high FOV will make anything in your hands look terrible. 

Most Godot users resort to having a separate Viewport overlaid on top of the original one. Not only is it cumbersome for users and error prone, it is really bad from a performance standpoint. 

`Z_CLIP_SCALE` allows users to pass a custom scaling factor to the shader that will scale the mesh towards the camera (so it doesn't clip other objects) without impacting how the mesh is sorted or shaded. This allows users to trivially place arms/objects/weapons in front of an FPS camera without a performance penalty and while maintaining normal lighting + shadows. 

`fov_override` is similar, it is implemented in the BaseMaterial3D only and can be used to override the FOV with which a an object is rendered. It is slightly more limited as it is designed for perspective cameras only, but it should work really well for most users. The most common feedback I received from users on an early draft of this PR was that they wanted the FOV to be specified in degrees and function as an override instead of being a scaling factor like I originally wrote (see https://x.com/solitary_ps/status/1801340839916147054)

`IN_SHADOW_PASS` has been asked for for a long time. I needed it to cleanly add `fov_override`, so I finally took the chance to expose it. Using the naming suggested here: https://github.com/godotengine/godot-proposals/issues/4443#issuecomment-1146079780. Its a bit awkward since the Forward+ renderer doesn't actually have a separate shader variant for shadows, so this has to use a uniform bool. But its not a big problem.

clayjohn committed:
Add new StandardMaterial properties to allow users to control FPS-style objects (hands, weapons, tools close to the camera)

Add new shader built in Z_CLIP_SCALE to easily adjust clipping distance to avoid clipping walls etc.

Add fov_override to StandardMaterial3D to easily have a custom FOV for FPS objects

Add IN_SHADOW_PASS built-in to shaders for tweaking materials without impacting shadow maps

Calinou reviewed (commented):
Tested locally on all rendering methods on Linux + NVIDIA, it works as expected.

**Testing project:** [test_perspective_scale.zip](https://github.com/user-attachments/files/15844415/test_perspective_scale.zip)

https://github.com/godotengine/godot/assets/180032/216dd593-e772-41ff-a8ef-1c71905e9ad8

Some feedback:

- Overall, this approach works well with weapon models made of multiple meshes, as done in the testing project above.

- The Viewport approach is still relevant if you wish to make your weapon model translucent while avoiding any sorting issues, including if multiple meshes overlap within the weapon model. For (mostly) opaque weapon models, this new approach is clearly better.

- Models with clip perspective can still cast and receive shadows, including from other objects in the scene. The perspective scale does not affect the position at which shadows are received, which I think makes sense for this use case.

- Screen-space effects no longer affect the model that uses clip scale and a low Z clip scale, but this is expected. We should make sure to document this, as I don't see a way to resolve this.

- **Z Clip Scale** and **Perspective Scale** should be hidden in the inspector if **Use Clip Scale** is unchecked.

- The perspective scale multiplier is inverted compared to what I'm typically expecting. Higher values result in a *lower* FOV, instead of a higher FOV. This might not be a problem, but I thought I'd mention it nonetheless.

- Meshes parented to the camera occasionally disappear with low perspective scale (= high FOV). See the example below with the weapon, which is made of multiple MeshInstance3D nodes:

https://github.com/godotengine/godot/assets/180032/def097f0-ee0e-49fb-b439-ea4090d8cc68

It might be possible to work around this one by setting a large custom AABB to the weapon model (as it's always visible anyway), but this isn't exposed in the Advanced Import Settings dialog yet.

thygrrr commented:
Does this work well with reverse Z?

Can this feature be adapted to work the other way around?
(to render distant backgrounds, e.g. space scenes, mountain ranges, etc.)

I think it probably can't, so I'd recommend implementing this with a remap instead of a simple scale.

Z_CLIP_MIN = 1
Z_CLIP_MAX = 0

with resulting Z thus being 
`z_result = original * (zc_min-zc_max) + zc_max)`
(possibly clamped to 0..1)

Changing Min would then be a simple scale.

Changing Max would be the "always in front of this max distance"

Changing Both would give fine control over either near-eye and far-field rendering.

clayjohn commented:
> Does this work intuitively with reverse Z?

Yes. Unless I am misunderstanding your question. Godot always uses reverse z and this works in Godot.

> Can this feature be adapted to work the other way around?

Not as it is currently written. I'm not sure that this simple z-scale trick would work for what you describe though. Keep in mind that anything that relies on depth will be broken when you manually adjust z-scale. 

This works for small objects like hands/objects because they don't contribute significantly to global illumination, but it totally breaks when multiple objects need to be sorted against each other (which is why this is a slider and not a toggle, I expect users will bring the scale up to 1 when the character needs to grab an object in the scene for example). 

For distant objects, you would have a big issue for transitions between normal objects and scaled objects and potentially between scaled objects. Accordingly a more complex solution would be needed for that case.

samanthar0se commented:
This would be a welcomed feature. Personally I'd advocate for `PERSPECTIVE_SCALE` to be represented as Degrees/FOV, but as long math to convert from FOV to whatever units are used, I think anything is fine.

Would it be possible to (optionally?) set that value globally? In the FPS use one will generally wan them all to present at the same FOV so having it synced globally would be ideal.

clayjohn commented:
This would be a welcomed feature. Personally I'd advocate for PERSPECTIVE_SCALE to be represented as Degrees/FOV, but as long math to convert from FOV to whatever units are used, I think anything is fine.

I think the consensus at this point is that this should be specified as an override in absolute degrees instead of a scale as it is now. I'm not sure about the math right now, but I am sure it is possible.

Would it be possible to (optionally?) set that value globally? In the FPS use one will generally wan them all to present at the same FOV so having it synced globally would be ideal.

I guess we could make the default FOV overridable in a project setting. But I'm not sure what exactly you are asking for. You can already set the FOV in the Camera3D node for all objects. This PR is for adding a scaling factor to specific meshes that you want to skew

samanthar0se commented:
> I guess we could make the default FOV overridable in a project setting. But I'm not sure what exactly you are asking for. You can already set the FOV in the Camera3D node for all objects. This PR is for adding a scaling factor to specific meshes that you want to skew

Sorry, I think I could have been more clear!

I think one of the most common use cases for this change would be (as you've called out in your original post) first person perspective hands/weapons. Typically in a FPS all the first person models are animated at a particular FOV in the modelling software and therefore the game generally wants to present all the first person models at that same FOV to maintain the scale of the hands/guns. Like you said, there may be instances where the scale needs to be manipulated for the hands to interact with the rest of the scene, but generally the baseline perspective scale for all first person models will be the same.

In that typical use case having a way to globally set a default `PERSPECTIVE_SCALE` for any material with `USE_CLIP_SCALE` unchecked would be help prevent bugs and simplify keeping all first person models at the same scale.

I'm not sure if this is possible, but I think an ideal solution would simply be a way to set the default value of `PERSPECTIVE_SCALE` in the project settings. Users who are using these settings in some other way can simply ignore that setting, and the variable can still be manipulated freely on each material if desired.

clayjohn commented:
Rebased and fixed merge conflicts. 

CC @Calinou Could you take another look? The current iteration should resolve all the awkward points from before

Calinou reviewed (approved):
Works great :slightly_smiling_face: I like the addition of `IN_SHADOW_PASS` too.

I think the new UX makes more sense overall. We lose out on some flexibility (orthogonal cameras, Keep Width aspect mode), but it's possible to emulate the Keep Width aspect mode with Keep Height using [some math on the FOV override](https://en.wikipedia.org/wiki/Field_of_view_in_video_games#Field_of_view_calculations).

**Updated testing project:** [test_perspective_scale_2.zip](https://github.com/user-attachments/files/20371648/test_perspective_scale_2.zip)

Repiteo commented:
Thanks!

catboomer1105 commented:
> **Updated testing project:** [test_perspective_scale_2.zip](https://github.com/user-attachments/files/20371648/test_perspective_scale_2.zip)
can't download the testing project, showing `Not Found`
